### ISSUE TITLE
cps version of `chain`

### ISSUE THREAD

**Issue by srijs** [fantasyland/fantasy-land#99]
In light of the coming support for tail-call optimisation in ES6, I'd like to propose the following addition to `Functor`:

    tailMap :: (a -> (b -> (f b -> c) -> c) -> c ) -> f a -> c

Such that:

    tailMap (\a g -> g (f a) id) == map f

Combined with `join`, it allows us to construct `tailChain`:

    var join = function (mm) {
      return mm.chain(function (m) { return m; });
    };

    var tailChain = function (m, fn) {
      return m.tailMap(function (a, wrap) {
        return fn(a, function (b, cb) {
          return wrap(b, function (mm) {
            return cb(join(mm));
          });
        });
      });
    };

Using continuation passing style, `tailChain` can take advantage of tail-call optimisation and eliminate the need for trampolining when sequencing a large number of actions.

**Comment by CrossEye** [comment]
Is this derivable in some way or are you suggesting that everyone who believesl their objects to be Functors would also need to implement `tailMap` ?

**Comment by srijs** [comment]
As demonstrated, `tailMap` is strictly more powerful than ap`, so deriving it is probably not an option.

If you don't want to change Functor, an addition could introduce a TailFunctor class, from which we can derive `TailChain`. Or it could simply be an optional member of Functor.

The idea would be that asynchronous code checks for `tailMap`, and falls back to `map` with a warning if it is not found.

I'm sure there are also other good ways to add it, I just want to raise awareness that the current Chain class does not have good real world characteristics, and propose a better alternative.

**Comment by bergus** [comment]
Uh, TCO is all fine and so, but what advantage does using CPS give us in the first place? Could you please show use a sample implementation of `map` vs `tailMap` on a data structure of your choice that demonstrates the difference?

I've tried and failed to make an implementation. The most difficult problem I encountered is that each `a` callback might use a different continuation, and you never really get back to `f b`. This makes `tailMap` look more like a `Traversable` or `Foldable` method, not a `Functor` one.

**Comment by CrossEye** [comment]
Ok, @bergus, how do people always beat me to these things by minutes?

I was going to ask exactly the same thing!

**Comment by srijs** [comment]
First of all, let me show you an implementation of the identity functor, with both map and tailMap:

    var Identity = function (a) {
      this.a = a;
    };

    Identity.prototype.map = function (f) {
      return new Identity(f(this.a));
    };

    Identity.prototype.tailMap = function (f) {
      return f(this.a, function (b, cont) {
        return cont(new Identity(b));
      }.bind(this));
    };

Now, that is not very exciting, but we can already see that for itself, the advantage `tailMap` provides over `map` is that `f` can compute `b` asynchronously.

Secondly, let's extend identity to a monad:

    Identity.of = function (a) {
      return new Identity(a);
    };

    Identity.prototype.join = function () {
      return new Identity(this.a.a);
    };

Based on that, we can now implement `chain` and `tailChain` on top of it:

    Identity.prototype.chain = function (f) {
      return this.join(this.map(f));
    };

    Identity.prototype.tailChain = function (f) {
      return this.tailMap(function (a, wrap) {
        return f(a, function (b, cont) {
          return wrap(b, function (wrapped) {
            return cont(wrapped.join());
          });
        });
      });
    };

So what is this all good for?

Say we have a function `transform` which performs some kind of expensive computation. Given the single-threaded nature of javascript, we want to run this function asynchronously. With `tailChain`, this is straight forward:

    id.tailChain(function (a, wrap) {
      return transform(a, function (b) {
        return wrap(b, function (id) {
          return id.tailChain(function (b, wrap) {
            return transform(b, function (c) {
              //...

With `chain`, not so much. We have to rely on some kind of trampoline, for example a free monad, which I will not illustrate here for the sake of brevity. The problem with that is that trampolining is complex and performs way worse than the simple CPS approach using `tailChain`.

There are other use-cases as well. For example, it is possible to emulate `do`-style monadic composition using ES6 generator functions. For that to work, you will need an interpreter function, which takes a generator function, reads all the `yield`ed values, and performs the monadic sequencing. In order to be able to implement this function in a functional style using recursion, you need to use `tailChain`, otherwise your implementation will not be tail-recursive, blowing the stack after some amount of chaining.

Regarding your point about the different continuations, @bergus, I believe it would be possible to move the continuation argument into the initial arguments, so that the signature for `tailMap` would look like:

    tailMap :: (a -> (b -> c) -> c) -> (f b -> d) -> f a -> d

This would enable implementations for functors which have multiple values to iterate over:

    var HomogenousTuple = function (a, b) {
      this.a = a;
      this.b = b;
    };

    HomogenousTuple.prototype.tailMap = function (f, cont) {
      return f(this.a, function (a) {
        return f(this.b, function (b) {
          return cont(new HomogenousTuple(a, b));
        });
      });
    };

**Comment by bergus** [comment]
@srijs 
> it would be better to move the continuation argument into the initial arguments. This would enable implementations for functors which have multiple values to iterate over

Ah, there we go. That's indeed the problem I had, as I tried to implement it for lists/arrays. Your original version wouldn't have worked for all Functors. Btw, I think that should read `… (f c -> d) …`.

>  it is possible to emulate `do`-style monadic composition using ES6 generator functions

That only works for structures that only contain up to one value as well.

> As most javascript APIs are inherently callback based

Well, we're trying to migrate all of them to promises :-) I'm still thinking about your other arguments. Those CPS monads look interesting, though I'm not sure whether it's just the CPS monad composed with a base one.

**Comment by srijs** [comment]
Well fwiw, anecdotally, most monads used in practice for control flow seem to be single-valued.

The tail-calling identity monad could just as well be written as a monad stack of `IdentityT Promise a`, and I am well aware of that. Using something like promises (or any other base monad) is precisely what I am trying to avoid, because of the aforementioned cost we pay for the trampolining ;)

Maybe I am in denial here, but I would like to think that there could be a way to write non-trivial control flow without resorting to the hammer that is promises. I am writing a run-loop for the mentioned ES6 generator technique right now, and it is literally not possible to write it for synchronous monads without `tailMap` (without blowing the stack _and_ still return synchronously).

**Comment by srijs** [comment]
After having some time to look for alternative solutions, I have found a different pattern that allows me to write said run-loop without recursion, but with a simple `while` construct instead. It might not be as general as the proposed `tailMap`, but it solves the issue for me.

In case you're interested, the bit that enables this is a type-class I call `TraverseNext`, which is effectively a specialised `Traversable` for the `Next` datatype. Let me demonstrate...

This is the signature of `Next`, which resembles the values that are returned from ES6 generators: 

    data Next a = Next { value :: a, done :: Boolean }

Now, the function every monad in the `Nextable` class implements is the following, which is a specialisation of `sequence`:

    sequenceNext :: m (Next a) -> Next (m a)

In practice, the implementation looks something like this:

    Identity.prototype.sequenceNext = function () {
      return {value: new Identity(this.a.value), done: this.a.done};
    };

    Writer.prototype.sequenceNext = function () {
      return {
        value: new Writer(this.w.concat(this.a.value.w), this.a.value.a),
        done: this.a.done
      };
    };

Using `sequenceNext`, we can derive `traverseNext` as:

    let traverseNext = function (f) { return this.map(f).sequenceNext(); };

Based on that, this is the simple run-loop I came up with:

    var run = function (m, gen) {
      let cur = m.of(null);
      while (true) {
        let next = cur.traverseNext(function (a) {
          return gen.next(a);
        });
        cur = next.value;
        if (next.done) {
          return cur;
        }
      }
    };

Why the special type-class? I am still trying to write a sensible `Applicative` instance for `Next`, but have failed so far. At this point, I am unsure whether `ap` can be written such that it has the correct behaviour to support my use-case, but I haven't stopped trying yet...
